### PR TITLE
add reset learning rate functionality

### PR DESCRIPTION
**Pull request by dimapihtar**
# What does this PR do ?

Adds reset learning rate functionality.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/core/optim/lr_scheduler.py [hunk 1]
```diff
@@ -270,6 +277,7 @@ def __init__(
         self.decay_steps = max_steps - (self.constant_steps + self.warmup_steps)
 
         self.min_lr = min_lr
+        self.first_step = True
```
Comment (yaoyu-33) [inline nemo/core/optim/lr_scheduler.py:280]: can you explain what this flag does?

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 2]
```diff
@@ -395,6 +395,11 @@ def __init__(self, cfg: DictConfig, trainer: Trainer):
 
         self.inference_params = None
 
+        # Reset learning rate params
+        self.if_init_step = True
```
Comment (yaoyu-33) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:399]: any difference between self.first_step vs. this ?

File: nemo/core/optim/lr_scheduler.py (hunk 1 above)
Comment (dimapihtar) [reply nemo/core/optim/lr_scheduler.py:280]: @yaoyu-33 it's needed to reset self.max_steps and self.decay_steps when reset_lr_steps is True.

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 3]
```diff
@@ -762,6 +767,15 @@ def training_step(self, dataloader_iter):
         if self.initialize_ub:
             self.initialize_ub_func()
 
+        # Reset learning rate
+        if self.if_init_step and self.reset_lr:
+            self._optimizer.param_groups[0]['lr'] = 0.0 if self.cfg.optim.sched.warmup_steps > 0 else self.cfg.optim.lr
```
Comment (yaoyu-33) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:772]: I assume this is the main place to reset lr? I am not super sure it will work for dist optimizer. 
Also what happens if there are two groups?

Comment (dimapihtar) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:772]: > I assume this is the main place to reset lr? I am not super sure it will work for dist optimizer. Also what happens if there are two groups?

yes, it works with dist optimizer.
yes, for now it only supports dist optimizer since the're going to use distributed one

File: nemo/core/optim/lr_scheduler.py (hunk 1 above)
Comment (dimapihtar) [reply nemo/core/optim/lr_scheduler.py:280]: we only need to reset them once at the beggining of the training

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py (hunk 2 above)
Comment (dimapihtar) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:399]: > any difference between self.first_step vs. this ?

hmm that's a good point
I think I can try to use one variable for both of places

**Comment by dimapihtar** [comment]
- Use case 1: `reset_lr=True, reset_lr_steps=False`
We want to reset_lr to init_lr (0 or optim.lr) in order to train it on another data and don't care about learning rate's max number of steps.

- Use case 2: `reset_lr=True, reset_lr_steps=False` 
We want to reset_lr and we also want to reach min_lr by the end of the training by using the same trainer.max_steps as for previous run of this checkpoint. So in this case we want to update lr's max_steps and decay_steps in the next way:
`max_steps = max_steps - consumed_steps,
decay_steps = decay_steps - consumed_steps`
where `conusmed_steps` is the number of steps already completed at the checkpoint.
Example:
image

**Review by yaoyu-33 — APPROVED** [review]
